### PR TITLE
ARM diags v2 integration

### PR DESCRIPTION
**Pull request by chengzhuzhang**
This PR integrates using in-situ measurements from ARM to compare with point grid model output. Features including:
- seasonal mean, annual cycle cycle of a set of atmospheric variables, including (surface temperature, precipitation, radiation fluxes, surface turbulent fluxes...)
- diurnal cycle of precipitation and cloud
- convection onset metrics

### CODE REVIEW

**Comment by chengzhuzhang** [comment]
@forsyth2 Hi Ryan, this PR is ready to review. The example output can be found here: https://portal.nersc.gov/cfs/e3sm/zhang40/tests/arm_diags_v2_m2o/viewer/arm_diags/index.html

@tomvothecoder will help apply the linters/formater tools after this PR
 
The one caveat is that i got an error in unit tests but not sure if this PR caused the problem. Any insights? 
```
======================================================================
ERROR: test_all_sets_and_all_seasons (__main__.TestRun)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "tests/test_run.py", line 54, in test_all_sets_and_all_seasons
    parameters = self.runner.get_final_parameters([self.core_param, ts_param, enso_param, streamflow_param])
  File "/Users/zhang40/miniconda3/envs/dev_dec2_2020_py3.8/lib/python3.8/site-packages/acme_diags/run.py", line 82, in get_final_parameters
    self.parser.check_values_of_params(final_params)
  File "/Users/zhang40/miniconda3/envs/dev_dec2_2020_py3.8/lib/python3.8/site-packages/cdp/cdp_parser.py", line 86, in check_values_of_params
    p.check_values()
  File "/Users/zhang40/miniconda3/envs/dev_dec2_2020_py3.8/lib/python3.8/site-packages/acme_diags/parameter/core_parameter.py", line 86, in check_values
    raise RuntimeError(msg)
RuntimeError: You need to define both the 'ref_start_yr' and 'ref_end_yr' parameter.

----------------------------------------------------------------------
```

File: acme_diags/viewer/arm_diags_viewer.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
+import os
+from .utils import add_header, h1_to_h3
+from .default_viewer import create_metadata
+from cdp.cdp_viewer import OutputViewer
+import csv
+#from varid_dict import varid_longname
```
Comment (forsyth2) [inline acme_diags/viewer/arm_diags_viewer.py:6]: Remove?

File: examples/tutorials/postprocessing_E3SM_data_for_single_sites.sh [hunk 2]
```diff
@@ -0,0 +1,82 @@
+import cdms2
+from pathlib import Path
+#from acme_diags.driver.utils import climo
```
Comment (forsyth2) [inline examples/tutorials/postprocessing_E3SM_data_for_single_sites.sh:3]: Remove?

**Review by forsyth2 — COMMENTED** [review]
@chengzhuzhang Overall looks good. I mostly made minor comments on things I didn't think the formatter would catch.

I did a high-level review of `acme_diags/driver/arm_diags_driver.py`, `acme_diags/plot/cartopy/arm_diags_plot.py`, and `examples/tutorials/postprocessing_E3SM_data_for_single_sites.sh`. If you'd like me to spend some time diving deeper into those, just let me know. I'd imagine if the plots look good and the tests pass, then everything is working as expected.

As for the one failing test, it looks like you'll need to add start and end years to the test https://github.com/E3SM-Project/e3sm_diags/blob/master/tests/test_run.py#L41. You'll probably also have to update the `all_seasons_count` in that test since these diags deal with sub-monthly plots.

**Comment by chengzhuzhang** [comment]
Thanks, @forsyth2! Regarding the comment below, in `core_parameter.py`,  it is by default:      
```
self.ref_timeseries_input = False
self.test_timeseries_input = False
```
So the check_value shouldn't apply. Also, I haven't changed `test_run.py` in this PR. Do you recall if you have seen this error before?

> As for the one failing test, it looks like you'll need to add start and end years to the test https://github.com/E3SM-Project/e3sm_diags/blob/master/tests/test_run.py#L41. You'll probably also have to update the `all_seasons_count` in that test since these diags deal with sub-monthly plots.

**Comment by forsyth2** [comment]
@chengzhuzhang Interesting.

https://travis-ci.org/github/E3SM-Project/e3sm_diags/builds/759450963 shows a different error:

```
======================================================================
ERROR: test_all_sets_and_all_seasons (test_run.TestRun)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/E3SM-Project/e3sm_diags/tests/test_run.py", line 54, in test_all_sets_and_all_seasons
    parameters = self.runner.get_final_parameters([self.core_param, ts_param, enso_param, streamflow_param])
  File "/home/travis/miniconda/envs/e3sm_diags/lib/python3.7/site-packages/acme_diags/run.py", line 56, in get_final_parameters
    other_params = self._get_other_diags(parameters[0].run_type)
  File "/home/travis/miniconda/envs/e3sm_diags/lib/python3.7/site-packages/acme_diags/run.py", line 242, in _get_other_diags
    default_diags_paths = [get_default_diags_path(set_name, run_type, False) for set_name in self.sets_to_run]
  File "/home/travis/miniconda/envs/e3sm_diags/lib/python3.7/site-packages/acme_diags/run.py", line 242, in <listcomp>
    default_diags_paths = [get_default_diags_path(set_name, run_type, False) for set_name in self.sets_to_run]
  File "/home/travis/miniconda/envs/e3sm_diags/lib/python3.7/site-packages/acme_diags/acme_diags_driver.py", line 50, in get_default_diags_path
    "Plotting via set '{}' not supported, file {} not installed".format(set_name, fnm))
RuntimeError: Plotting via set 'arm_diags' not supported, file arm_diags_model_vs_obs.cfg not installed
```

I get the above error when testing locally as well.

File: acme_diags/plot/cartopy/arm_diags_plot.py [hunk 3]
```diff
@@ -0,0 +1,436 @@
+import os
+import matplotlib
+matplotlib.use('agg')
```
[the remark this answers is not in the document]
Comment (tomvothecoder) [reply acme_diags/plot/cartopy/arm_diags_plot.py:3]: Thanks for the heads up. This will be on my radar once I rebase.

**Comment by chengzhuzhang** [comment]
Thanks! This is helpful! Looks like the .cfg files were not pushed in...

File: acme_diags/derivations/acme.py [hunk 4]
```diff
@@ -56,6 +56,9 @@ def convert_units(var, target_units):
     elif var.units == 'mb/day':
         var = var
         var.units = target_units
+    elif var.id == 'prw' and  var.units == 'cm':
+        var = var * 10.0 #convert from 'cm' to 'kg/m2' or 'mm'
```
[the remark this answers is not in the document]
Comment (chengzhuzhang) [reply acme_diags/derivations/acme.py:60]: Yes, with the approximation: water density = 1000 kg/m³. A kg of water over a square meter equivalent to 1mm depth of rain.

File: acme_diags/driver/arm_diags_driver.py [hunk 5]
```diff
@@ -0,0 +1,377 @@
...
+                parameter.viewer_descr[var] = getattr(test, 'long_name', var)
+                # Get the name of the data, appended with the years averaged.
+                parameter.test_name_yrs = utils.general.get_name_and_yrs(parameter, test_data)
+                parameter.var_name = getattr(test, 'long_name', var)
+                parameter.var_units = getattr(test, 'units', var)
+
+                refs = []
+
+                #for ref_name in ref_names:    
+                #setattr(parameter, 'ref_name', ref_name)
```
[the remark this answers is not in the document]
Comment (chengzhuzhang) [reply acme_diags/driver/arm_diags_driver.py:74]: done

File: acme_diags/driver/arm_diags_driver.py [hunk 6]
```diff
@@ -0,0 +1,377 @@
...
+                        msg = 'Diurnal cycle of {} at Site: {} is not supported yet'.format(region, var)
+                        raise RuntimeError(msg)
+                    else:
+                        ref_file_name = 'sgparmdiagsmondiurnalC1.c1.nc'
+ 
+                        ref_file = os.path.join(ref_path,ref_file_name)
+                        ref_data = cdms2.open(ref_file)
+                        
+                        if var == 'PRECT':
+                            ref = ref_data('pr')*3600.*24
```
[the remark this answers is not in the document]
Comment (chengzhuzhang) [reply acme_diags/driver/arm_diags_driver.py:86]: done

File: acme_diags/driver/arm_diags_driver.py [hunk 7]
```diff
@@ -0,0 +1,377 @@
...
+                arm_diags_plot.plot_diurnal_cycle(var, vars_to_data[season], parameter)
+
+
+    return parameter
+
+
+def run_diag_diurnal_cycle_zt(parameter):
+    variables = parameter.variables
+    regions = parameter.regions
+    #ref_names = parameter.ref_names
```
[the remark this answers is not in the document]
Comment (chengzhuzhang) [reply acme_diags/driver/arm_diags_driver.py:130]: done

File: acme_diags/driver/arm_diags_driver.py [hunk 8]
```diff
@@ -0,0 +1,377 @@
...
+                    parameter, ignore_container=True), parameter.output_file + '.json')
+                print('Metrics saved in: ' + fnm)
+   
+                arm_diags_plot.plot_diurnal_cycle(var, vars_to_data[season], parameter)
+
+
+    return parameter
+
+
+def run_diag_diurnal_cycle_zt(parameter):
```
[the remark this answers is not in the document]
Comment (chengzhuzhang) [reply acme_diags/driver/arm_diags_driver.py:127]: zt indicates fields with both z axis and time axis.

**Comment by forsyth2** [comment]
@chengzhuzhang I just reran the tests with your changes and was able to reproduce the error. I turned off the `timeseries_input` options in `acme_diags/parameter/arm_diags_parameter.py`. That caused `test_run.py TestRun.test_all_sets_and_all_seasons` to pass but `test_diags.py TestAllSets.test_image_count` to fail. I wonder if the `timeseries_input` parameters are getting overwritten in the wrong places?

**Comment by chengzhuzhang** [comment]
@forsyth2 Thanks. Yes, i also suspect that `timeseries_input` parameters are not defined correctly. I'm looking into it...

**Comment by chengzhuzhang** [comment]
@forsyth2 Hey Ryan, the latest commit fixed the test. It appears that the error was caused by not specifying sets_to_run. I'm still puzzled why this problem was not shown earlier though...If things look okay, i will squash and merge. So that @tomvothecoder  can move the other PRs forward.

File: acme_diags/driver/utils/dataset.py [hunk 9]
```diff
@@ -228,8 +240,11 @@ def get_start_and_end_years(self):
             else:
                 start_yr = getattr(self.parameters, 'test_start_yr')
                 end_yr = getattr(self.parameters, 'test_end_yr')
+        
+        if self.parameters.sets[0] in ['diurnal_cycle', 'arm_diags']:
+            sub_monthly = True
 
-        return start_yr, end_yr
+        return start_yr, end_yr, sub_monthly
```
[the remark this answers is not in the document]
Comment (forsyth2) [reply acme_diags/driver/utils/dataset.py:247]: I did a `git grep "get_start_and_end_years("` and it looks like all occurrences have been updated.

**Comment by forsyth2** [comment]
@chengzhuzhang Glad you found the issue in the test, but I agree that is strange it hasn't been causing errors before. I did a quick re-review of change files and I think it should be good to merge.

**Pull request merged**